---
title: "eqtl results exploration"
##output: html_notebook
---
```{r}
suppressMessages({
    library(arrow)
    library(data.table)
    library(BiocParallel)
    library(qs2)
    library(SummarizedExperiment)
    library(stringi)
    library(GenomicRanges)
    library(here)
  })
here::i_am('.git/HEAD')
datadir=here("processed-data", "rds")
## refdata folder/symlink to CHESS Brain data for now
refdir=here("processed-data", "ref")
```


```{r}
## load all rowRanges 
granges <- qs_read("granges.qs2")
g2sym <- setNames(granges$gene_name, granges$gene_id)
# load degs adust SPD
deg_spd_adj <- fread(file.path(datadir, '10_dx_deg_adjust_spd/dx-deg_PRECAST07.csv'), data.table = FALSE)
rownames(deg_spd_adj) <- deg_spd_adj$V1
setnames(deg_spd_adj, 'V1', 'fid')

fgwasflt <- file.path(refdir, 'SCZ_GWAS_flt_1e-6_R.8.tab.gz')
if (file.exists(fgwasflt)) {
    gwas <- fread(fgwasflt) # 30205 variants
} else {
  stop("file not found: ", fgwasflt) ## are you sure you want to rebuild this?
  allgwas <- fread(file.path(refdir, 'SCZ_GWAS_hg38_euro.tab.gz')) # 7659767
  ##GWAS human standard: < 5e-8
  ##         suggestive: < 1e-6
  gwas <- allgwas[p<1e-6 & !is.na(pos) & impinfo>=0.8] ## 30205 variants
  fwrite(gwas, fgwasflt, sep='\t', quote=FALSE)
}
gwas_set <- unique(gwas[p<5e-8]$variant_id) ## GWAS standard risk threshold
gwas1e6_set <- unique(gwas$variant_id) ## loose p < 1e-6
degs <- subset(deg_spd_adj, fdr_scz<=0.05)
```

```{r}
pats <- c('\\.map_cis\\.tab\\.gz$', '\\.map_independent\\.txt\\.gz$',
           '\\.DxINT\\.cis_qtl_top_assoc\\.txt\\.gz$')
npats <- c('map', 'indep', 'dxint')
cisres <- list()
for (i in seq_along(pats)) {
  fcis <- list.files(path='tqtl_out', pattern=pats[i], recursive = TRUE, full.names = TRUE)
  names(fcis) <- sub("/", "", stri_match_first_regex(fcis, '/(\\w+)\\.gene')[,2])
  cisres[[npats[i]]] <- lapply(fcis, fread)
}
ds2name <- setNames( c("SpD01-WMtz", "SpD02-L3/4", "SpD03-L6",  "SpD04-WM", "SpD05-L5", 
       "SpD06-L2/3","SpD07-L1", "Neun", "Vasc", "PNN", "Neuropil"),  c(paste0('spd0',1:7), "neun", "vasc", "pnn", "neuropil"))
```

```{r}
## tag dataset to each data.table before merging
tag_ll <- function(dt_list) {
  rbindlist(lapply(names(dt_list), function(nm) {
    x <- copy(dt_list[[nm]])
    x[, ds:=nm]
    x[, gid := phenotype_id]
    ## add gtype and symbol
    x[, symbol :=g2sym[phenotype_id]]
    x[, dge:=as.integer(phenotype_id %chin% degs$fid)]
    # convert dataset
    x[, ds := as.factor(ds)]
    x[, gwas := as.integer(variant_id %chin% gwas_set)]
    x
  }))
}

map_all   <- tag_ll(cisres$map)     # has qval
indep_all <- tag_ll(cisres$indep)   # has rank
dx_all    <- tag_ll(cisres$dxint)   # has pval_gi, b_g, b_gi, pval_adj_bh

#dedup_by_ds <- function(dt) unique(dt, by = c("ds","phenotype_id"))

```

```{r discovery and analysis}
get_qmap  <- function(emap, q) {
  emap[qval < q, .( 
    n_eGenes = uniqueN(phenotype_id),                          # unique eGenes
    n_GWAS      = uniqueN(phenotype_id[gwas == 1]),          # eGenes whose lead is in GWAS (5e-8)
    GWAS = paste(sort(unique(symbol[gwas==1])), collapse=", "),
    n_DEG   = uniqueN(gid[dge   == 1]),
    DEGS = paste(sort(unique(symbol[dge==1])), collapse=", "),
    n_DEG_GWAS =  uniqueN(phenotype_id[dge==1 & gwas==1]),
    DEG_GWAS = paste(sort(unique(symbol[dge==1 & gwas==1])), collapse=", ")
  ),
  by = .(ds)
][order(ds)]
}
## eGene cis_map counts at eQTL FDR 5%
mapsum05 <- get_qmap(map_all, 0.05)
mapsum05$ds <- ds2name[mapsum05$ds]
mapsum05
```

> mapsum05
            ds n_eGenes n_GWAS                           GWAS n_DEG n_DEG_GWAS   DEG_GWAS
        <char>    <int>  <int>                         <char> <int>      <int>     <char>
 1: SpD01-WMtz      134      2              LRRC37A2, ZSCAN31     7          0           
 2: SpD02-L3/4      161      4 HLA-DMA, MAPK3, VARS2, ZSCAN31     9          1      MAPK3
 3:   SpD03-L6        5      0                                    1          0           
 4:   SpD04-WM       10      0                                    1          0           
 5:   SpD05-L5      179      2              LRRC37A2, ZSCAN31     8          0           
 6: SpD06-L2/3       48      2                HLA-C, LRRC37A2     7          0           
 7:   SpD07-L1        2      0                                    1          0           
 8:       Neun       61      1                       LRRC37A2     4          0           
 9:       Vasc       61      2           KANSL1-AS1, LRRC37A2     6          1 KANSL1-AS1
10:        PNN       12      1                     KANSL1-AS1     3          1 KANSL1-AS1
11:   Neuropil        3      0                                    1          0           


```{r plot barchart}
library(ggplot2)

eqtl_barchart  <- function(mapsum, indep=FALSE, fill=NULL) {
  # pal_lib <- c(PolyA = "#82BEEE", RiboZ = "#FFBE6A") 
  dt <- copy(mapsum)
  dt[, context  := as.factor(ds)] ## set levels explictly here
  dt[, lab_above := sprintf("GWAS: %d\nDEG: %d\nDEG∩GWAS: %d", n_GWAS, n_DEG, n_DEG_GWAS)]
  if (indep) {
    ytitle <- "Total independent cis-eQTL signals (p_perm < 0.05)"
    title  <- "Independent cis-eQTLs by context"
  } else {
    ytitle <- "eGenes discovered (lead cis-eQTLs, q < 0.05)"
    title  <- "Lead cis-eQTLs per eGene by context"
  }
  #mapsum[, lab_count := sprintf("%d", n_eGenes)]
  min_count <- min(dt[n_eGenes>0]$n_eGenes)
  # label positions inside each bar (relative to its own height)
  dt[, y_top  := n_eGenes]                    # top-of-bar (for the big count)
  dt[, y_gwas := n_eGenes - min_count * 0.08]             # just below the top
  dt[, y_de   := n_eGenes - min_count * 0.08]             # a bit further down, top-aligned

  # two rows per bar: GWAS and Other; stack within each bar
  bar_df <- rbindlist(list(
    dt[, .(context, category="GWAS (p<5e-8)", count=n_GWAS)],
    dt[, .(context, category="Other",         count=n_eGenes - n_GWAS)]
  ))
  if (is.null(fill)) {
     p <- ggplot(dt, aes(x=context, y=n_eGenes, fill=context)) +
       geom_col(width=0.7, color="grey30", linewidth=0.2)
  } else {
     p <- ggplot(dt, aes(x=context, y=n_eGenes)) +
       geom_col(width=0.7, color="grey30", linewidth=0.2, fill=fill)
  }   
  p <- p +
    #facet_grid(level ~ coding, scales="free_y") +
    # big total on top of each bar
    geom_text(aes(y = y_top, label = n_eGenes), vjust = -0.3,
              fontface = "bold", size = 3.3, color = "black") +
    # inside-bar annotations
    geom_text(aes(y = y_top, label = lab_above), vjust = -0.5,
              color = "black", size = 3) +
    # scale_fill_manual(values = pal_lib, guide = "none") +
    scale_y_continuous(expand = expansion(mult = c(0, 0.20))) +
    coord_cartesian(clip = "off") +
    labs(x = NULL, y = ytitle, title = title) +
    theme_bw(base_size = 14) +
    theme(
      panel.grid.minor = element_blank(),
      axis.ticks.x     = element_blank(),
      strip.background = element_rect(fill = "grey95", color = "grey80")
    )
 p
}
print(eqtl_barchart(mapsum05[grepl('^SpD', ds)]))
print(eqtl_barchart(mapsum05[!grepl('^SpD', ds)], fill="#A0A0A0"))
```

```{r independent signals}
## independent signals per feature (rank is 1..k per phenotype)

alpha <- 0.1  # primary FDR line for discovery - typically 0.05
## parent eFeature q-values from map_all
parent_q <- map_all[, .(phenotype_id, ds, qval_parent = qval)]

## Note that independent signals are already filtered by pval_beta<=0.05 in the parent eFeature
## since cis.map_independent() only tests features with pval_beta<=0.05 from cis.map_cis
### -- join and gate using pval_perm, which is slightly more conservative than pval_beta
indep_f <- merge(indep_all, parent_q, by=c("phenotype_id","ds"))
indep_f <- indep_f[qval_parent < alpha & pval_perm < 0.05]

indep_var <- indep_f[, .(
  n_eGenes    = .N,                             # total independent variants (despite the name)
  n_GWAS      = sum(gwas),                      # variants in GWAS 5e-8
  n_DEG  = uniqueN(gid[dge == 1]),
), by = .(ds)]

# per-feature summary (avoids double counting signals)
indep_feat <- indep_f[, .(n_signals   = max(rank),
  any_gwas    = as.integer(any(gwas == 1)),
  any_gwas1e6 = as.integer(any(gwas1e6 == 1)),
  any_dge     = as.integer(any(dge == 1)),
  any_dge10   = as.integer(any(dge10 == 1)),
  any_dte     = as.integer(any(dte == 1)),
  any_dtu     = as.integer(any(dtu == 1))
), by = .(lib, level, coding, gid, phenotype_id)]

## overall signal architecture: how many independent cis-eQTL signals each feature has
sig_arch <- indep_feat[, .(
  median = as.numeric(median(n_signals)),
  mean   = as.numeric(mean(n_signals)),
  p95    = as.numeric(quantile(n_signals, 0.95, names = FALSE))
), by = .(lib, level)]

k95_tbl <- indep_feat[, .(p95_nsig = as.numeric(quantile(n_signals, 0.95)),
                          n_DEG_GWAS = uniqueN(phenotype_id[any_dge==1 & any_gwas==1])
                  ), by = .(ds)]
plotdf <- merge(indep_var, k95_tbl, by = c(ds), all.x = TRUE)


## summarize by lib, level, coding, with proper unique counting
indep_summary <- indep_feat[, .( n_feat = as.integer(.N),  # unique eFeatures with ≥1 indep signal
      med_nsig   = as.numeric(median(n_signals)),
      # feature-level overlaps (no double counting across signals)
      n_gwas     = as.integer(sum(any_gwas)),
      # transcript-level overlaps: count unique transcripts
      n_DEG   = as.integer(uniqueN(gid[any_dge   == 1])),
      n_DEG_GWAS = uniqueN(phenotype_id[any_dge==1 & any_gwas==1])
  ),
  by = .(ds)
][order(ds)]

#print(eqtl_barchart(indep_summary, indep=TRUE))
print(eqtl_barchart(plotdf, indep=TRUE))
```


```{r interaction summary}
## choose thresholds
#thr_primary <- 0.10  # primary significance threshold
thr_primary <- 0.25  # exploratory
dx_dups <- dx_all[, .N, by=.(ds, phenotype_id)][N > 1]
if (nrow(dx_dups) > 0) {
  ## there are duplicate entries in dx_all for some features; keeping the first only
  #print(dx_dups)
  dx_all <- dx_all[order(pval_gi)][, .SD[1], by=.(ds, phenotype_id)]
}
dx_best_qc <- dx_all[af >= 0.05 & af <= 0.95 & ma_count >= 20]
## primary and exploratory thresholds
dx_primary  <- dx_best_qc[pval_adj_bh <= thr_primary]
#dx_explore  <- dx_best_qc[pval_adj_bh <= 0.25]

summ_dx <- function(dt) dt[, .(
  n_feat        = .N,
  prop_pos_bgi  = mean(b_gi > 0),
  med_abs_bgi   = median(abs(b_gi)),
  n_GWAS        = sum(gwas),
  n_DEG    = uniqueN(gid[dge == 1]),
), by=.(ds)][order(ds)]

dxsum10 <- summ_dx(dx_primary)

```